### DISCUSSION TITLE
Batch updating pictures via API

### DISCUSSION THREAD

**Discussion by picklerick804** [photoprism/photoprism#3884] [General]
Hi team!

Being this my first post, THANK YOU for this project! It's super cool!

I am using it to manage around 500k pictures coming from different sources, one of my requirements is to enrich the pictures metadata. To achieve that I wrote my - rudimental - client library that is performing a `PUT /api/v1/photos/:uid`

Things were good until i passed the mark of 450k pictures and now every PUT operation is really slow/heavy.

My brief investigation shows the following:

- MariaDB is working really hard (high CPU usage)
- it seems the `UpdatePhoto` calls the `UpdateClientConfig` which invoke the `ClientUser`
- There there is an query that summarises the `photos` table content:
`
SELECT 
    SUM(photo_type = 'video' AND photo_quality > -1 AND photo_private = 0) AS videos, 
    SUM(photo_type = 'live' AND photo_quality > -1 AND photo_private = 0) AS live, 
    SUM(photo_quality = -1) AS hidden, 
    SUM(photo_type NOT IN ('live', 'video') AND photo_quality > -1 AND photo_private = 0) AS photos, 
    SUM(photo_quality BETWEEN 0 AND 2) AS review, 
    SUM(photo_favorite = 1 AND photo_private = 0 AND photo_quality > -1) AS favorites, 
    SUM(photo_private = 1 AND photo_quality > -1) AS private 
FROM `photos`  WHERE 
    (photos.id NOT IN (SELECT photo_id FROM files WHERE file_primary = 1 AND (file_missing = 1 OR file_error <> ''))) AND (deleted_at IS NULL) 
LIMIT 1
`
- Running the query directly into MariaDB console shows it is really slow (~10sec)

**The environment**
- The official docker compose photprism + mariadb
- VM hosted on i5 intel cpu and 4GB ram (this is the only service running there)
- MariaDB storage is the local NVMe

**Questions**
- Do you see any issue with my setup?
- What about making the `UpdateClientConfig` non blocking, would it break anything ?
- Is there a better way to update pictures metadata from an external service?

Thanks!

**Comment by lastzero** [comment]
For best performance (and if you trust the client), you can perform the changes directly in MariaDB without using the Update Photo API. We will add a batch edit API, but it's not ready yet.

**Reply by picklerick804** [reply]
Thanks @lastzero ,

Good to hear a batch api is coming.  I am going to look into your suggestion. I mainly update tags and time stamps, do you see anything I should worry about ?

**Reply by lastzero** [reply]
You might have to run the `photoprism optimize` command after performing changes so that these are reflected e.g. in automatically generated titles and albums.

**Reply by picklerick804** [reply]
Allright,

I swapped my REST client with a DB-Direct one. My main goal is to update keywords (adding/removing) What are the involved tables?

- `keywords` holds the keyword definition. I am fetching the id if existent or adding a new one
- `photos_keywords` holds the M-to-M relations between `photos` and `keywords` right?
- Do I need to update the `details` table too ?
- I've seen some cache references in the code, do I need to invalidate it somehow?

I really appreciate your help on this. 🙏 
Best

**Reply by lastzero** [reply]
It will be enough if you populate any of the fields shown here with the additional keywords:
https://github.com/photoprism/photoprism/blob/052fe82126988b3d1ec95fe2e04b2cd61bf10a78/internal/entity/photo.go#L453-L464

The background worker will then automatically update the indexes. In order to accelerate this, you can e.g. set the checked_at column to NULL in the photos table:
https://github.com/photoprism/photoprism/blob/052fe82126988b3d1ec95fe2e04b2cd61bf10a78/internal/query/photo.go#L112-L113
